### PR TITLE
feat: Introduce Dedicated Istio Gateway Secret

### LINKED ISSUE #1890: feat: [Zero-Downtime] - safe migration
### Description

In order to introduce the [zero-downtime](https://github.com/kyma-project/lifecycle-manager/issues/1430) procedure, we need a safe migration path.
By "safe migration path" I understand a setup, where a "revert" to the old behavior is as simple as possible - in case there is a bug in the new solution or it is not working as expected for whatever reason.
In particular, the revert should be as simple as switching some Lifecycle-Manager flags (runtime arguments) - the less, the better.

#### An Idea for how this could work:
The current solution uses root certificate secret as the Istio-Gateway secret directly.
Once the root cert is rotated, the LM code deletes the client secrets (that are based on the "previous" root), causing cert-manager to renew them.
The new ("zero-downtime") solution uses a dedicated secret for the Istio-Gateway, managed entirely by Lifecycle-Manager.
This new secret of course has a different name from the "root" secret, that is still managed and rotated by the cert-manager.
The dedicated secret decouples Istio-Gateway from changes to the root certificate - it is Lifecycle-Manager that decides when to propagate the changes.
How can we revert from the new solution to the old one?
If the secret name used for Istio-Gateway is different in the old (current) and new (future) solution, then in order to switch back we would have to change the Helm Charts, or at least the entry in the `values.yaml`. In addition we would probably need to deploy a different LM version - the one with the "old" logic. But then we're reverting the Lifecycle-Manager version, along with all the other features, security fixes etc.

To improve the situation, we should:
- have a LM version that is capable of running either the old or the new code - depending on some feature flag.
- the Istio Gateway secret name should be **the same** in both scenarios, to avoid manual tweaks in the Helm charts.

The first requirement is relatively easy - we just need to extract the relevant cert-management logic to a component and then provide two different implementations: the old one (current) and the new one. And we need a flag to decide which component should be actually used at runtime.

The second requirement is more tricky. In order to make it work, we should change the Lifecycle-Manager in the following way (it's just an idea, maybe it can be done in a simpler way):
- The root certificate is no longer directly used to configure Istio Gateway
- We introduce our own, managed secret that is a copy of the Istio-Gatewy secret.
- There is a new "agent" that actively syncs the root secret to the copy.
- Minimal changes in the code are required - The reconciliation logic no longer inspects the timestamps on the root secret, it inspects the created copy instead
- Kustomize configuration of the Istio Gateway (and Helm charts) are changed so that the new secret is used instead of the "root" one. 

By introducing this solution ("safe-migration"), we achieve the following:
- Both the "safe-migration" and "final" solutions use **the same secret name** for the Istio-Gateway. Hence, when reverting, there's no need to change anything in the Helm charts.
- The revert can be accomplished by switching a single boolean flag on the Lifecycle-Manager, like: `--cert-mangament-legacy=true`, assuming no additional configuration flags are required.
- Don't worry about the "copy" or the new "agent" - in the final solution these also exists. And this issue introduces just a temporary solution. Once the final solution works, we'll remove the support for the "old" logic entirely.

Implementation notes for the syncing "agent" - for lack of a better name, let's call it: "Istio Gateway Secret Manager"
- What we really need is something that watches the root secret and acts upon changes on this object
- So it can be a single goroutine, a mini-controller, etc. Maybe even something else - a little research is required.
- See that although the secret name is the same, the data in the secret differs. The "safe-migration" solution makes IstioGatewaySecret just a copy of the root secret, while the final solution uses a **modified** (newRoot+oldRoot) values for the caBundle in the IstioGatewaySecret.

See also: https://github.com/kyma-project/lifecycle-manager/issues/1506
### Reasons

Safe migration path - ability to revert **fast** and with minimal risk of doing it wrong - in case of troubles.
### Implementation notes:
- introduce a new mechanism for copying the secret. Can be a single goroutine with a watcher/informer. It should have a low-frequency polling in case k8s event is lost (AFAIK k8s events are not 100% reliable)
- modify the condition upon which the skr watcher client secrets are removed: https://github.com/kyma-project/lifecycle-manager/blob/f0ac9bb8aa0c8770409b1123e2c4fbf4d75980f4/pkg/watcher/certificate_manager.go#L299
- if necessary, consider adding some annotation on the Istio Gateway Secret (IGS) object to make "rotation detection" in the code easy.

### Acceptance Criteria

- [x] Double-check the proposed solution if it's not missing anything important
- [ ] Verify with security team if the LM-managed secret is OK (we need this anyway for the "new solution")
- [x] introduce a secret manager component that does a copy
- [x] change the [condition for watcher client secret removal](https://github.com/kyma-project/lifecycle-manager/blob/main/pkg/watcher/certificate_manager.go#L299) so that it involves updates to the Istio Gataway Secret (which should be an up-to-date copy of the Root Secret)

### Feature Testing

_No response_

### Testing approach

_No response_

### Attachments

_No response_

### Related Issues:
https://github.com/kyma-project/lifecycle-manager/issues/1430

### PR DESCRIPTION
**Pull request by LeelaChacha**
Closes #1890

### CODE REVIEW

File: pkg/zerodw/gateway.go [hunk 1]
```diff
@@ -0,0 +1,137 @@
+package zerodw
```
Comment (ruanxin) [inline pkg/zerodw/gateway.go:1]: please rename this package to something readable, I don't know what "dw" means

File: config/watcher/gateway.yaml [hunk 2]
```diff
@@ -20,5 +20,5 @@ spec:
         number: 443
         protocol: HTTPS
       tls:
-        credentialName: klm-watcher
+        credentialName: istio-gateway-secret
```
Comment (ruanxin) [inline config/watcher/gateway.yaml:23]: Can you shortly explain why we have to change the credentailName here? also in this PR, you name it [gateway-secret](https://github.com/kyma-project/lifecycle-manager/pull/1871/files#diff-076f5cb5879121781dc0b5aaf99c461555cbab02b46a9b13b94482242f0aa985R23), and we previously have issue regarding alignment of resource name, https://github.com/kyma-project/lifecycle-manager/pull/1721#issuecomment-2256185803, please also have a reference to that decision for this renaming.

Comment (ruanxin) [reply config/watcher/gateway.yaml:23]: If it make sense to have a short meeting to sync the decision, please don't hesitate bring it to team, you can sync this with @Tomasz-Smelcerz-SAP first.

Comment (LeelaChacha) [reply config/watcher/gateway.yaml:23]: would `klm-istio-gateway` be more fitting then?

File: pkg/watcher/certificate_manager.go [hunk 3]
```diff
@@ -318,6 +292,14 @@ func (c *CertificateManager) RemoveSecretAfterCARotated(ctx context.Context, kym
 	return nil
 }
 
-func certificateRenewalTimePassed(certStatus certmanagerv1.CertificateStatus) bool {
-	return certStatus.RenewalTime.Before(&(apimetav1.Time{Time: time.Now()}))
+func isGatewaySecretNewerThanWatcherCert(gatewaySecret *apicorev1.Secret, watcherSecret *apicorev1.Secret) bool {
```
Comment (ruanxin) [inline pkg/watcher/certificate_manager.go:295]: I would like to see a unit test for this function

File: pkg/watcher/skr_webhook_manifest_manager.go [hunk 4]
```diff
@@ -87,9 +85,17 @@ func (m *SKRWebhookManifestManager) Install(ctx context.Context, kyma *v1beta2.K
 		return fmt.Errorf("failed to get skrContext: %w", err)
 	}
 
+	_, err = getCertificateSecret(ctx, m.kcpClient, client.ObjectKey{
```
Comment (ruanxin) [inline pkg/watcher/skr_webhook_manifest_manager.go:88]: you can save a few lines if put this into if condition as `if _, err := getCertificateSecret(...); err != nil`

File: pkg/watcher/certificate_manager.go (hunk 3 above)
Comment (LeelaChacha) [reply pkg/watcher/certificate_manager.go:295]: good point!

File: pkg/watcher/skr_webhook_manifest_manager.go [hunk 5]
```diff
@@ -237,13 +246,24 @@ func (m *SKRWebhookManifestManager) getUnstructuredResourcesConfig(ctx context.C
...
 		skrWatcherImage: m.config.SkrWatcherImage,
-		caCert:          tlsSecret.Data[caCertKey],
+		caCert:          gatewaySecret.Data[caCertKey],
 		tlsCert:         tlsSecret.Data[tlsCertKey],
 		tlsKey:          tlsSecret.Data[tlsPrivateKeyKey],
 		remoteNs:        remoteNs,
 	}, nil
 }
 
+func getCertificateSecret(ctx context.Context, clnt client.Client, objKey client.ObjectKey) (*apicorev1.Secret, error) {
```
Comment (ruanxin) [inline pkg/watcher/skr_webhook_manifest_manager.go:256]: all of objKey are: `Namespace: m.certificateConfig.IstioNamespace, Name: gatewaysecret.GatewaySecretName,`, right? why not design this function belongs to `CertificateManager` as `func (c *CertificateManager) getCertificateSecret(ctx context.Context)`

Comment (ruanxin) [reply pkg/watcher/skr_webhook_manifest_manager.go:256]: I expected a function comment to explain the purpose of this secret.

File: cmd/main.go [hunk 6]
```diff
@@ -201,12 +204,21 @@ func setupManager(flagVar *flags.FlagVar, cacheOptions cache.Options, scheme *ma
 	go cleanupStoredVersions(flagVar.DropCrdStoredVersionMap, mgr, setupLog)
 	go scheduleMetricsCleanup(kymaMetrics, flagVar.MetricsCleanupIntervalInMinutes, mgr, setupLog)
 
+	setupIstioGatewaySecretRotation(config, kcpClient, setupLog)
```
Comment (ruanxin) [inline cmd/main.go:207]: I would prefer put the goruntine declaration here so that aligns with above method and has a clear view that all of them are go rountine.

File: pkg/gatewaysecret/secret.go [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+	err := sm.kcpClient.Get(ctx, objKey, secret)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get secret %s: %w", objKey.Name, err)
+	}
+
+	return secret, nil
+}
+
+// isNotFound returns true if the error is a NotFound error.
+func isNotFound(err error) bool {
```
Comment (ruanxin) [inline pkg/gatewaysecret/secret.go:33]: we have utility isNotFound https://github.com/kyma-project/lifecycle-manager/blob/0d91106f187024f74f65846cf78222d8ad471079/pkg/util/error.go#L15

reuse that please

File: pkg/gatewaysecret/handler.go [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+
+func (gsh *GatewaySecretHandler) handleExisting(rootSecret *apicorev1.Secret, gwSecret *apicorev1.Secret) error {
+	caCert := certmanagerv1.Certificate{}
+	if err := gsh.kcpClient.Get(context.Background(),
+		client.ObjectKey{Namespace: istioNamespace, Name: kcpCACertName},
+		&caCert); err != nil {
+		return fmt.Errorf("failed to get CA certificate: %w", err)
+	}
+
+	if gwSecretLastModifiedAtValue, ok := gwSecret.Annotations[LastModifiedAtAnnotation]; ok {
```
Comment (ruanxin) [inline pkg/gatewaysecret/handler.go:65]: I would expect this get last modified timestamp is a function from `secretManager`

File: pkg/gatewaysecret/handler.go [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+
+const (
+	GatewaySecretName = "istio-gateway-secret"
+	kcpRootSecretName = "klm-watcher"
+	kcpCACertName     = "klm-watcher-serving"
+	istioNamespace    = flags.DefaultIstioNamespace
+)
+
+type GatewaySecretHandler struct {
+	*secretManager
```
Comment (ruanxin) [inline pkg/gatewaysecret/handler.go:28]: Why need this wrapper struct only for `secretManager`, can you just use secretManager do all the thing?

File: pkg/gatewaysecret/handler.go [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+		Data: map[string][]byte{
+			"tls.crt": rootSecret.Data["tls.crt"],
+			"tls.key": rootSecret.Data["tls.key"],
+			"ca.crt":  rootSecret.Data["ca.crt"],
+		},
+	}
+	return gwSecret
+}
+
+func WatchChangesOnRootCertificate(clientset *kubernetes.Clientset, gatewaySecretHandler *GatewaySecretHandler,
```
Comment (ruanxin) [inline pkg/gatewaysecret/handler.go:105]: I expect the scenario you designed for this function fully tested in an dedicated e2e test, for example, the root cert get updated, test your watched one should get new secret synced, right?

File: pkg/gatewaysecret/handler.go [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+
+		switch event.Type {
+		case watch.Added:
+			fallthrough
+		case watch.Modified:
+			err := gatewaySecretHandler.ManageGatewaySecret(rootCASecret)
+			if err != nil {
+				log.Error(err, "unable to manage istio gateway secret")
+			}
+		case watch.Deleted:
```
Comment (ruanxin) [inline pkg/gatewaysecret/handler.go:127]: if it's deleted, we don't clean up the secret?

File: tests/e2e/ca_certificate_rotation_test.go [hunk 12]
```diff
@@ -34,17 +33,21 @@ var _ = Describe("CA Certificate Rotation", Ordered, func() {
 			Namespace: RemoteNamespace,
 		}
 		It("Then KCP TLS Certificate is removed", func() {
-			timeNow := &apimetav1.Time{Time: time.Now()}
-			expectedLogMessage := "CA Certificate was rotated, removing certificate"
+			var err error
+			namespacedSecretName := types.NamespacedName{
+				Name:      watcher.ResolveTLSCertName(kyma.Name),
+				Namespace: "istio-system",
```
Comment (ruanxin) [inline tests/e2e/ca_certificate_rotation_test.go:39]: think about put this `istio-system` as a const and shared within e2e test.

File: pkg/gatewaysecret/handler.go (hunk 11 above)
Comment (LeelaChacha) [reply pkg/gatewaysecret/handler.go:127]: thats is an undefined state. it should never be deleted AFAIK. Even if there is a workflow in the cefrt manager to delete it, it would be added again, that's when we'll tigger the update again. Note that secret being deleted doesn't mean that the copied gateway secret is invalid. It expires only at the `notAfter` date

File: pkg/gatewaysecret/handler.go (hunk 10 above)
Comment (LeelaChacha) [reply pkg/gatewaysecret/handler.go:105]: but that is already being tested in the ca rotation test. The root cert is set to rotate every 1 minute and we check if the skr cert is rotated. so we are are also checking the rotation of the gateway by proxy. I would suggest that we explicitly check the gateway secret in same test, rather than a dedicated test. WDZT? @ruanxin

File: pkg/gatewaysecret/handler.go (hunk 11 above)
Comment (ruanxin) [reply pkg/gatewaysecret/handler.go:127]: then please write a code comment here to explain this case.

File: pkg/gatewaysecret/handler.go [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+}
+
+func (gsh *GatewaySecretHandler) handleNonExisting(rootSecret *apicorev1.Secret) error {
+	gwSecret := gsh.newGatewaySecret(rootSecret)
+	return gsh.create(context.Background(), gwSecret)
+}
+
+func (gsh *GatewaySecretHandler) handleExisting(rootSecret *apicorev1.Secret, gwSecret *apicorev1.Secret) error {
+	caCert := certmanagerv1.Certificate{}
+	if err := gsh.kcpClient.Get(context.Background(),
```
Comment (ruanxin) [inline pkg/gatewaysecret/handler.go:59]: we should not use Background context, check the function comment, this context will never cancelled.

File: config/watcher/gateway.yaml (hunk 2 above)
Comment (ruanxin) [reply config/watcher/gateway.yaml:23]: I would vote for `klm-istio-gateway`

File: pkg/watcher/skr_webhook_manifest_manager.go (hunk 5 above)
Comment (LeelaChacha) [reply pkg/watcher/skr_webhook_manifest_manager.go:256]: This function is used in 5 places. Sometimes for the gateway secret and sometimes for the watcher secret. Hence was made generic

File: config/watcher/gateway.yaml (hunk 2 above)
Comment (LeelaChacha) [reply config/watcher/gateway.yaml:23]: I'll use that for now, but let's discuss it briefly after a daily.

File: pkg/watcher/skr_webhook_manifest_manager.go (hunk 5 above)
Comment (LeelaChacha) [reply pkg/watcher/skr_webhook_manifest_manager.go:255]: makes sense!

File: pkg/watcher/certificate_manager_test.go [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+					ObjectMeta: apimetav1.ObjectMeta{
+						Annotations: map[string]string{
+							"lastModifiedAt": "2024-11-01T00:00:00Z",
+						},
+					},
+				},
+				watcherSecret: &apicorev1.Secret{
+					ObjectMeta: apimetav1.ObjectMeta{
+						CreationTimestamp: apimetav1.Time{
+							Time: time.Date(2024, 11, 0o1, 0, 0, 5, 0, time.UTC),
```
Comment (ruanxin) [inline pkg/watcher/certificate_manager_test.go:36]: why use `0o1`? can't be `0`?

File: tests/e2e/rbac_privileges_test.go [hunk 15]
```diff
@@ -178,7 +178,7 @@ var _ = Describe("RBAC Privileges", func() {
 				{
 					APIGroups: []string{""},
 					Resources: []string{"secrets"},
-					Verbs:     []string{"list", "watch", "create", "delete"},
+					Verbs:     []string{"list", "watch", "create", "delete", "update"},
```
Comment (ruanxin) [inline tests/e2e/rbac_privileges_test.go:181]: please also update this permission in `management-plane-charts` lifecycle-manager/files/roles/klm-watcher-certmanager-role.yaml before release

File: cmd/main.go (hunk 6 above)
Comment (ruanxin) [reply cmd/main.go:207]: what I mean is just like this 
```
	go cleanupStoredVersions(flagVar.DropCrdStoredVersionMap, mgr, setupLog)
	go scheduleMetricsCleanup(kymaMetrics, flagVar.MetricsCleanupIntervalInMinutes, mgr, setupLog)

	go setupIstioGatewaySecretRotation(config, kcpClient, setupLog)
}

func setupIstioGatewaySecretRotation(config *rest.Config, kcpClient *remote.ConfigAndClient, setupLog logr.Logger) {
	kcpClientset := kubernetes.NewForConfigOrDie(config)
	gatewaySecretHandler := gatewaysecret.NewGatewaySecretHandler(kcpClient)
	gatewaysecret.WatchChangesOnRootCertificate(kcpClientset, gatewaySecretHandler, setupLog)
}
```

File: tests/e2e/suite_test.go [hunk 16]
```diff
@@ -32,6 +32,7 @@ import (
 const (
 	kcpConfigEnvVar, skrConfigEnvVar = "KCP_KUBECONFIG", "SKR_KUBECONFIG"
 	clientQPS, clientBurst           = 1000, 2000
+	istioNamespace                   = "istio-system"
```
Comment (ruanxin) [inline tests/e2e/suite_test.go:35]: actually I have a better place to declaire namespace used for tests, https://github.com/kyma-project/lifecycle-manager/pull/2030/files#diff-5dab9c6f73d5f7fde3503baa47c3f8ef0528addf1143961dac081556181d7ebbR15-R16, if you PR is merged after that one, please update accordingly, otherwise, do this in another PR.

File: pkg/testutils/watcher.go [hunk 17]
```diff
@@ -104,3 +151,30 @@ func GetCACertificate(ctx context.Context, namespacedCertName types.NamespacedNa
...
+	if err != nil {
+		return fmt.Errorf("failed to get gateway secret %w", err)
+	}
+
+	currentTime, err := gatewaysecret.GetValidLastModifiedAt(gwSecret)
+	if err != nil {
+		return fmt.Errorf("failed to get last modified time %w", err)
+	}
+
+	if currentTime != oldTime {
```
Comment (ruanxin) [inline pkg/testutils/watcher.go:176]: I think to only compare different is not solid enough, currentTime should larger than oldTime.

File: config/manager/kustomization.yaml [hunk 18]
```diff
@@ -10,5 +10,5 @@ generatorOptions:
 
 images:
 - name: controller
-  newName: europe-docker.pkg.dev/kyma-project/prod/lifecycle-manager
-  newTag: latest
+  newName: k3d-kcp-registry.localhost:5000/lifecycle-manager
```
Comment (Tomasz-Smelcerz-SAP) [inline config/manager/kustomization.yaml:13]: looks like a leftover from local testing?

File: pkg/watcher/skr_webhook_manifest_manager.go [hunk 19]
```diff
@@ -87,9 +85,13 @@ func (m *SKRWebhookManifestManager) Install(ctx context.Context, kyma *v1beta2.K
 		return fmt.Errorf("failed to get skrContext: %w", err)
 	}
 
+	if _, err = gatewaysecret.GetGatewaySecret(ctx, m.kcpClient); err != nil {
```
Comment (Tomasz-Smelcerz-SAP) [inline pkg/watcher/skr_webhook_manifest_manager.go:88]: Isn't that redundant? Later on, this value is required anyway so we make another call like this.

File: pkg/watcher/certificate_manager.go [hunk 20]
```diff
@@ -279,35 +277,17 @@ func (e *CertificateNotReadyError) Error() string {
...
 func (c *CertificateManager) RemoveSecretAfterCARotated(ctx context.Context, kymaObjKey client.ObjectKey) error {
-	caCertificateStatus, err := c.GetCACertificateStatus(ctx)
+	gatewaySecret, err := gatewaysecret.GetGatewaySecret(ctx, c.kcpClient)
 	if err != nil {
-		return fmt.Errorf("error while fetching CA Certificate: %w", err)
+		return err
 	}
-
-	certSecret, err := c.getCertificateSecret(ctx)
+	watcherCert, err := c.getCertificateSecret(ctx)
```
Comment (Tomasz-Smelcerz-SAP) [inline pkg/watcher/certificate_manager.go:285]: I think it should be renamed to `watcherSecret`, as it is a secret object.

File: pkg/watcher/certificate_manager.go [hunk 21]
```diff
@@ -279,35 +277,17 @@ func (e *CertificateNotReadyError) Error() string {
...
-		c.caCertCache.SetCACertToCache(caCert)
-		return caCert.Status, nil
-	}
-
-	return cachedCertStatus, nil
-}
-
 func (c *CertificateManager) RemoveSecretAfterCARotated(ctx context.Context, kymaObjKey client.ObjectKey) error {
-	caCertificateStatus, err := c.GetCACertificateStatus(ctx)
+	gatewaySecret, err := gatewaysecret.GetGatewaySecret(ctx, c.kcpClient)
```
Comment (Tomasz-Smelcerz-SAP) [inline pkg/watcher/certificate_manager.go:281]: This object will be fetched very frequently. In the previous implementation we used an explicit cache. Is this call cached properly?

File: pkg/gatewaysecret/handler.go [hunk 22]
```diff
@@ -0,0 +1,182 @@
+package gatewaysecret
```
Comment (Tomasz-Smelcerz-SAP) [inline pkg/gatewaysecret/handler.go:1]: There are no unit tests in this package.
As we try to increase our code quality, it would be nice to provide some tests here.
A hint for what can be tested: I'd include the secret-watching logic, all time comparisons here and also that the "lastModifiedAt" annotation is always present on the secret.
Remember that the unit tests should not just be written to "test a scenario": They should be written with an intention to document the expected behavior of the unit (in our case, the unit is usually a single function/method)

File: pkg/gatewaysecret/handler.go [hunk 23]
```diff
@@ -0,0 +1,182 @@
...
+	return time.Time{}, errCouldNotGetLastModifiedAt
+}
+
+func WatchChangesOnRootCertificate(clientset *kubernetes.Clientset, gatewaySecretHandler *GatewaySecretHandler,
+	log logr.Logger,
+) {
+	ctx, cancel := context.WithCancel(context.TODO())
+	defer cancel()
+
+	secretWatch, err := clientset.CoreV1().Secrets(istioNamespace).Watch(ctx, apimetav1.ListOptions{
```
Comment (Tomasz-Smelcerz-SAP) [inline pkg/gatewaysecret/handler.go:151]: In this point a strong compile-time dependency on k8s library code is introduced.
It makes it impossible to properly unit-test our logic contained in the `for event := ...` loop.

File: pkg/watcher/certificate_manager.go (hunk 21 above)
Comment (LeelaChacha) [reply pkg/watcher/certificate_manager.go:281]: all secret objects are [cached by the Kyma reconciler](https://github.com/kyma-project/lifecycle-manager/blob/39caf9bb6be863211bde44c17b392cd296bfabb2/internal/controller/kyma/setup.go#L67)

File: pkg/watcher/skr_webhook_manifest_manager.go (hunk 19 above)
Comment (LeelaChacha) [reply pkg/watcher/skr_webhook_manifest_manager.go:88]: Yes, it was a part of the issue that the watcher secret is only created if the gateway secret is ready.

Later on, we fetch the gateway secret to check if the watcher secret needs to be updated; if an error occurs there, we don't update the watcher secret, but it will be created.
Therefore, stopping the watcher's secret from being created altogether when the gateway is not ready makes sense and is not redundant.

File: pkg/gatewaysecret/handler.go (hunk 22 above)
Comment (LeelaChacha) [reply pkg/gatewaysecret/handler.go:1]: I have added extensive unit tests for all the functions in `handler.go`. The struct now has a dependency injection to make it more testable. PTAL since I am not sure if my testing and mocking approach is a standard one.

File: pkg/watcher/skr_webhook_manifest_manager.go (hunk 19 above)
Comment (Tomasz-Smelcerz-SAP) [reply pkg/watcher/skr_webhook_manifest_manager.go:88]: ```Later on, we fetch the gateway secret to check if the watcher secret needs to be updated;```
I still don't get it. We fetch the secret in two places in this file: line 88 and 235. This places are quite "close" to each other with respect to actual execution. Can't we use the object fetched in line 88 also in line 235?
I mean, if there is a justified likelyhood that this secret somehow changes between lines 88 and 235, then it's fine to fetch it again.
Otherwise it is a code smell: Somehow the current code structure makes it hard for us to fetch the object just once and then use it.
If you don't know how to fix it in the current codebase (Or you don't want to do it in this PR), I'd suggest to create an issue dedicated to refactoring of this "double lookup".

File: pkg/gatewaysecret/handler_test.go [hunk 24]
```diff
@@ -0,0 +1,351 @@
...
+						NotBefore: &apimetav1.Time{
+							Time: time.Date(2024, 11, 1, 0, 0, 5, 0, time.UTC),
+						},
+					},
+				},
+			},
+			want: true,
+		},
+		{
+			name: "gateway secret is newer but has parsing error for lastModifiedAt",
```
Comment (Tomasz-Smelcerz-SAP) [inline pkg/gatewaysecret/handler_test.go:150]: When there is a parsing error, we don't know if it's newer. Maybe `gateway secret has parsing error for lastModifiedAt` is enough. But that is a very important test case, it documents that when there is a problem with `lastModifiedAt`, we consider this as "requires update", which makes sense, as the update should "fix" the invalid condition.

File: pkg/gatewaysecret/handler.go [hunk 25]
```diff
@@ -0,0 +1,142 @@
...
+		return nil
+	}
+
+	return gsh.Update(ctx, gwSecret)
+}
+
+func CopyRootSecretDataIntoGatewaySecret(gwSecret *apicorev1.Secret, caCert certmanagerv1.Certificate,
+	rootSecret *apicorev1.Secret,
+) bool {
+	if !GatewaySecretRequiresUpdate(gwSecret, caCert) {
```
Comment (Tomasz-Smelcerz-SAP) [inline pkg/gatewaysecret/handler.go:75]: Following the "Single Responsibility Principle" I don't think putting this condition here is a good choice.
First, it complicates the unit test: You have to setup some timestamps just to test if "copy" works...
Second, it complicates the signature of this function, `caCert` is used only to test if copying is required.
I would move this check to the place where `CopyRootSecretDataIntoGatewaySecret` is invoked (i.e. the body of the `handleExisting` method)

File: pkg/gatewaysecret/handler_test.go [hunk 26]
```diff
@@ -0,0 +1,351 @@
...
+		},
+		Data: map[string][]byte{
+			"tls.crt": []byte(oldTLSCertValue),
+			"tls.key": []byte(oldTLSKeyValue),
+			"ca.crt":  []byte(oldCACertValue),
+		},
+	}
+
+	// Newer than gateway secret
+	caCert := certmanagerv1.Certificate{
```
Comment (Tomasz-Smelcerz-SAP) [inline pkg/gatewaysecret/handler_test.go:198]: I think the `CopyRootSecretDataIntoGatewaySecret` function should be testable without having to fiddle with timestamps. Please see my other comment.

File: pkg/gatewaysecret/handler_test.go [hunk 27]
```diff
@@ -0,0 +1,356 @@
...
+	}
+	createFunc := func(ctx context.Context, gwSecret *apicorev1.Secret) error {
+		require.Equal(t, newTLSCertValue, string(gwSecret.Data["tls.crt"]))
+		require.Equal(t, newTLSKeyValue, string(gwSecret.Data["tls.key"]))
+		require.Equal(t, newCACertValue, string(gwSecret.Data["ca.crt"]))
+
+		return nil
+	}
+	mockManager := MockSecretManager{findGatewaySecretFunc: findGatewaySecretFunc, createFunc: createFunc}
+	handler := gatewaysecret.GatewaySecretHandler{SecretManager: mockManager}
```
Comment (Tomasz-Smelcerz-SAP) [inline pkg/gatewaysecret/handler_test.go:262]: I think this test is too complex to set up; it indicates a "too much coupling" condition.
Take a look at the event loop. You don't need entire `*GatewaySecretHandler` instance there. You use just one function. The event loop test should just test if this function is called with correct arguments for certain event types, and that the error is correrctly reported.
In that case you only need to mock one function - the one that is invoked in the event loop.

File: config/manager/kustomization.yaml (hunk 18 above)
Comment (Tomasz-Smelcerz-SAP) [reply config/manager/kustomization.yaml:13]: Once more :)

**Review by Tomasz-Smelcerz-SAP — APPROVED** [review]

**Review by Tomasz-Smelcerz-SAP — APPROVED** [review]
